docs/design/extending-api.md

```diff
@@ -0,0 +1,227 @@
+versions:
+  - name: v1
+    apiGroup: stable
+  - name: v2
+    apiGroup: experimental
+```
+
+Then the API server will program in two new RESTful resource paths:
+   * `/thirdparty/example.com/stable/v1/namespaces/<namespace>/crontab/...`
+   * `/thirdparty/example.com/experimental/v2/namespaces/<namespace>/crontab/...`
+
+
+Now that this schema has been created, a user can `POST`:
+
+```json
+{
+   "metadata": {
+     "name": "my-new-cron-object"
+   },
+   "apiVersion": "v1",
+   "kind": "example.com/ChronTab",
+   "cronSpec": "* * * * /5",
+   "image":     "my-awesome-chron-image"
+}
+```
+
+to: `/third-party/example.com/stable/v1/namespaces/default/crontab/my-new-cron-object`
+
+and the corresponding data will be stored into etcd by the APIServer, so that when the user issues:
+
+```
+GET /third-party/example.com/stable/v1/namespaces/default/crontab/my-new-cron-object`
+```
+
+And when they do that, they will get back the same data, but with additional Kubernetes metadata
+(e.g. `resourceVersion`, `createdTimestamp`) filled in.
+
+Likewise, to list all resources, a user can issue:
+
+```
+GET /third-party/example.com/stable/v1/namespaces/default/crontab
+```
+
+and get back:
+
+```json
+{
+   "apiVersion": "v1",
+   "kind": "example.com/ChronTabList",
+   "items": [
+     {
+       "metadata": {
+       "name": "my-new-cron-object"
+     },
+     "apiVersion": "v1",
+     "kind": "example.com/ChronTab",
+     "cronSpec": "* * * * /5",
+     "image":     "my-awesome-chron-image"
+    }
+   ]
+}
+```
+
+Because all objects are expected to contain standard Kubernetes metdata fileds, these
+list operations can also use `Label` queries to filter requests down to specific subsets.
+
+Likewise, clients can use watch endpoints to watch for changes to stored objects.
+
+
+## Storage
+
+In order to store custom user data in a versioned fashion inside of etcd, we need to also introduce a
+`Codec`-compatible object for persistent storage in etcd.  This object is `ThirdPartyResourceData` and it contains:
+   * Standard API Metadata
+   * `Data`: The raw JSON data for this custom object.
+
+### Storage key specification
+
+Each custom object stored by the API server needs a custom key in storage, this is described below:
+
+#### Definitions
+
+   * `resource-namespace` : the namespace of the particular resource that is being stored
+   * `resource-name`: the name of the particular resource being stored
+   * `third-party-resource-namespace`: the namespace of the `ThirdPartyResource` resource that represents the type for the specific instance being stored.
+   * `third-party-resource-name`: the name of the `ThirdPartyResource` resource that represents the type for the specific instance being stored.
+
+#### Key
+
+Given the definitions above, the key for a specific third-party object is:
+
+```
+${standard-k8s-prefix}/third-party-objects/${third-party-resource-namespace}/${third-party-resource-name}/${resource-namespace}/${resource-name}
```

Comment:
I don't think we should allow the same version number for different groups, that's bound to get super confusing.

But yes, the plan is to store whatever version the user supplies, and (for now) give back an error if the user asks for a 'v2' object that is stored in 'v1', once we implement HTTP plugins for APIs, then we can have a conversion plugin over HTTP.